### DISCUSSION TITLE
Looking for source code for 2 of the claimed features

### DISCUSSION THREAD

**Discussion by IannoIITR** [IBM/aihwkit#470] [Q&A]
I was going through the examples and I was unable to find the code for the following features described in the readme.

1. Forward pass output-referred noise and device fluctuations, as well as adjustable ADC and DAC discretization and bounds
2. Stochastic update pulse trains for rows and columns with finite weight update size per pulse coincidence
could you please share the functions implementing the above features?

**Comment by maljoras** [comment]
Hi @IannoIITR,
Forward pass output referred noise and ADC / DAC discretization is enabled and adjusted by setting the `rpuconfig.forward.out_noise` and `rpuconfig.forward.inp/out_res` properties of the class `IOParameters` as described in the documentation (see e.g. [here](https://aihwkit.readthedocs.io/en/latest/api/aihwkit.simulator.configs.utils.html?highlight=out_noise#aihwkit.simulator.configs.utils.IOParameters.out_res))

Device-to-device fluctuations are modeled an various levels, one is by adding read noise for each mat-vec, there are different noise types for that, see the options [here](https://github.com/IBM/aihwkit/blob/master/src/aihwkit/simulator/configs/enums.py#L102)     

There are a lot of parameters for device-to-device variations during  analog training, where a pulse update is made. These are handled by the different devices, for instance you can look at the example 10 which plots pulse responses to different device presets (see [here](https://github.com/IBM/aihwkit/blob/master/examples/10_plot_presets.py#L39)).   There are many parameter options in the devices which typical contain the string "dtod" for device-to-device variations. 

I am not sure why you are interested in the specific source implementations, but certainly all source code is available on this github. The functionality you are referring to are mostly handled with optimized C++/CUDA kernels and thus not part of the python code. They are in the RPUCuda library, for instance,   [this function](https://github.com/IBM/aihwkit/blob/master/src/rpucuda/sparse_bit_line_maker.cpp#L153) generates stochastic pulse trains for CPU and the [this file](https://github.com/IBM/aihwkit/blob/master/src/rpucuda/cuda/bit_line_maker.cu) contains many optimized kernels for generating compressed bit trains which are used by the pulse updater for CUDA.
